src/TraceEvent/EventPipe/NetTraceFormat_v6.md

```diff
@@ -0,0 +1,288 @@
+The intent of the MajorVersion and MinorVersion fields is that MajorVersion increments indicate breaking changes and MinorVersion increments indicate back-compatible changes.
+
+After the stream header comes the Trace object (which is now renamed to Trace block)
+
+#### Naming adjustment object -> block
+
+The FastSerialization abstraction refered to all the top level items in the file after the stream header as 'objects' and then some types of individual objects were EventBlock, MetadataBlock, etc. Starting in V6 FastSerialization is gone so we are also getting rid of the 'object' terminology and just calling them all "blocks" going forward. This part is purely a descriptive change and has no impact on the format.
+
+#### Updated block header
+
+In V5 an object header consisted of:
+- BeginPrivateObject Tag
+- SERIALIZED TYPE
+  - BeginPrivateObject Tag
+  - NullReference Tag
+  - version (int)
+  - minimum reader version (int)
+  - type name (string)
+  - EndPrivateObject Tag
+
+This header is 16 bytes + a variable length UTF8 string indicating the actual type of the object. The length of the object content was not included in the header. The EventBlock, MetadataBlock, and SPBlock objects included a BlockSize field as the first 4 bytes of the object content and the Trace object had fixed size.
+
+In V6 the object header is replaced by a simplified block header:
+
+```
+struct BlockHeader
+{
+  uint24 BlockSize
+  uint8 BlockKind
+}
+```
+
+Readers may want to read the BlockHeader as a single little endian 4 byte integer X where BlockSize = X & 0xFFFFFF and BlockKind = X >> 24.
+
+
+The type name strings used in the object header type name field map to new BlockKind values:
+
+```
+enum BlockKind
+{
+  EndOfStream = 0 // this is a new block type described later
+  Trace = 1,
+  Event = 2,
+  Metadata = 3,
+  Stack = 4,
+  SequencePoint = 5
+}
+```
+
+In V6 the EventBlock, MetadataBlock, StackBlock and SPBlock no longer have the initial BlockSize and padding fields. The value previously stored in the BlockSize object field is now in the BlockHeader. The semantics of the BlockSize header field are very similar to before, it represents the number of bytes in the remainder of the block. Specifically if file offset X points to the start of a block, then X + sizeof(BlockHeader) + BlockSize points to the start of the next block.
+
+Readers should skip over any blocks with an unrecognized BlockKind. New blocks might be introduced in minor version updates to store new types of data and this is expected to be a back-compatible change because old readers skip over it.
+
+
+#### Removed block footer
+
+In V5 all objects ended with a trailing EndPrivateObject tag byte. In V6 this trailing byte serves no purpose and is removed.
+
+#### Changed end of stream marker
+
+In V5 the file ended with a NullReferenceTag byte. In V6 the file ends with a new EndOfStream block. This block consists only of a BlockHeader and has BlockSize=0.
+
+## Multi-process support
```

Comment:
I understand that this is just making the format multi-process aware, but how do we envision this to be used? Currently each EventPipe session owns the whole stream, and there is plenty of state included serializing the stream, like caching stack traces, sequence point management, compression etc. If we would like to include multiple processes in the same session stream, then I would assume each EventPipe session would just store the raw events into memory, similar to how buffer manager currently does it internally and then have an external trace process that could trace multiple dotnet processes at the same time, meaning that the trace process will take care of most of the serialization currently done in the per session EventPipe serialization layer?

Reply:
The background context here is that we are planning to create a Linux profiling tool (based on perf_events and user_events) that will be doing full machine profiling. We wanted an open data format to write the trace info into that PerfView already had decent support for so we thought nettrace would be a good option. In this scenario it won't be EventPipe creating the data at all. At the moment there is no plan for EventPipe to create a multi-process trace.

Reply:
OK, so that means that this new profiling tool will handle the full serialization of the nettrace file and rely on rutnime user_events to get the data out.